KAFKA-4266: ReassignPartitionsClusterTest - Ensure ZK publication completed before start

benstopford opened the pull request:
Increase the reliability of the one temporal comparison in ReassignPartitionsClusterTest by imposing a delay after ZK is updated. This should be more reliable than just increasing the amount of data.

This relates to a previous PR: https://github.com/apache/kafka/pull/1982

benstopford committed:
KAFKA-4266 - check if each broker has had the throttled replicas updated before proceeding with the test (where applicable)

benstopford committed:
KAFKA-4266 - revert prev change as it's not really useful

benstopford committed:
KAFKA-4266 - add delay after ZK update for throttle config.

benstopford committed:
KAFKA-4266 - formatting

ijuma commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -322,18 +326,19 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, proposedAssignment: Map[TopicA
+        configs.put(DynamicConfig.Broker.FollowerReplicationThrottledRateProp, throttle.value.toString)
         AdminUtils.changeBrokerConfig(zkUtils, Seq(id), configs)
       }
+      Thread.sleep(throttle.postUpdateDelay)
```
Can you please elaborate a little on which scenario is made more reliable by this sleep?

benstopford replied:
If the rebalance kicks in before the throttle is set data isn't replicated. The current solution is to just make the test run for longer, with more data, so it averages out. 

It's in response to this commit: https://github.com/apache/kafka/commit/5e4600ae469559b9cdb36c17efb90cc6b6388826

silpamittapalli commented:
We too are experiencing fragility in ReassignPartitionsClusterTest and this PR seems to address the fragile test. Are there any plans of merging this PR?

ijuma commented:
@silpamittapalli, have you verified that this fixes the issues you are seeing?

silpamittapalli commented:
Yes, it addresses the fragile test @ijuma

benstopford commented:
happy to rebase if @ijuma has time to review?

ijuma commented:
@benstopford, yes, I can review it tomorrow.

benstopford commented:
@ijuma rebased dude

ijuma commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -24,11 +24,15 @@ import org.I0Itec.zkclient.exception.ZkNodeExistsException
 
 object ReassignPartitionsCommand extends Logging {
 
+  case class Throttle (value: Long, postUpdateDelay: Int = 0)
```
A minor nit: could we take a function `postUpdate: () => Unit` and then do the sleep in the test? That seems a bit nicer than having the sleep in the command. Apart from that, this looks OK, I could not think of a better way to detect that the throttling information had propagated.

benstopford replied:
Good call. Done

ijuma commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -322,18 +326,19 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, proposedAssignment: Map[TopicA
+        configs.put(DynamicConfig.Broker.LeaderReplicationThrottledRateProp, throttle.value.toString)
+        configs.put(DynamicConfig.Broker.FollowerReplicationThrottledRateProp, throttle.value.toString)
         admin.changeBrokerConfig(zkUtils, Seq(id), configs)
       }
```
One other question, should the delay be here or after `assignThrottledReplicas`?

benstopford replied:
Agreed - switched the order so all config is updated before the post update action. Considered putting a post update action into each method but it seems like more noise so I think this is the best balance.

benstopford committed:
KAFKA-4266:
- Pulled out sleep into test using a post update action.
- Reordered config updates so sleep is at the end.

benstopford commented:
@ijuma changes made as requested.

ijuma commented on core/src/main/scala/kafka/admin/Throttle.scala:
```diff
@@ -0,0 +1,28 @@
+
+package kafka.admin
+
+case class Throttle (value: Long, action: PostUpdateAction = NoPostUpdateAction)
```
Is there a reason why you used a `PostUpdateAction` class instead of just a function? e.g.:

```scala
case class Throttle(value: Long, postUpdateAction: () => Unit = () => ())
```

benstopford replied:
Laziness (i needed a class at an earlier point).

benstopford committed:
KAFKA-4266: Refactored out redundant class

benstopford commented:
@ijuma  Removed the redundant class. Thanks for the review.

ijuma commented:
Thanks, LGTM. I did a few minor tweaks before merging.
